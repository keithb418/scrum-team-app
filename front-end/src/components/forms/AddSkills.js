import React, { Component } from "react";
import { FormGroup, FormControl, InputGroup, ControlLabel } from "react-bootstrap";
import FontAwesome from "react-fontawesome";
import PropTypes from "prop-types";

export default class AddSkills extends Component {
  constructor (props) {
    super(props);
    this.state = { 
      skills: this.props.skills || [],
      skill: '' 
    };
  }

  onAdd = () => { 
    this.setState(({skills, skill}) => ({ skills: skills.concat(skill) }));
    this.props.onChange(this.state.skills);
    this.resetInput();
  }

  onRemove = (e) => {
    let skillToRemove = e.target.parentNode.innerText;
    this.setState(({skills}) => ({ skills: skills.filter(skill => skill != skillToRemove) }));
    this.props.onChange(this.state.skills);
  }

  getInput = (e) => {
    const value = e.target.value;
    const name = e.target.name;

    this.setState({
      [name]: value
    });
  }

  resetInput = () => {
    this.setState(() => ({ skill: '' }));
  }

  render () {
    let skillsToAdd = this.state.skills;
    return (
      <div className="add-skills">
        <FormGroup controlId={this.props.id} style={{ width: "200px" }}>
          <ControlLabel>Skills</ControlLabel>
          <InputGroup>
            <FormControl name="skill" type="text" placeholder="Add Skills" value={this.state.skill} onChange={this.getInput}/>
            <InputGroup.Addon>
              <FontAwesome name="plus-circle" className="left-spacer add-skills-icon" style={{ cursor: "pointer" }} onClick={this.onAdd} />
            </InputGroup.Addon>
          </InputGroup>
        </FormGroup>

        {skillsToAdd.map((skill, index) => {
          return (
            <div style={{ display: "inline-block", marginRight: "15px" }} className="skill" key={index}>{skill}
              <FontAwesome name="minus-circle" style={{ cursor: "pointer" }} onClick={this.onRemove} />
            </div>
          );
        })}

      </div>
    );
  }
}

AddSkills.propTypes = {
  id: PropTypes.string,
  onChange: PropTypes.func
};