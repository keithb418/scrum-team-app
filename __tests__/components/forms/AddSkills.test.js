import React from "react";
import { shallow } from "enzyme";
import AddSkills from "../../../front-end/src/components/forms/AddSkills";
let skills, wrapper, resetInput, onChange;

beforeEach(() => {
  skills = ["react", "redux", "es6", "backbone"]
  onChange = jest.fn();
  resetInput = jest.fn();
  wrapper = shallow(
    <AddSkills
        id="add-skills"
        skills={skills}
        onChange={onChange}/>
  );
});

test('should render AddSkills component correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('handles changine input for skill', () => {
  wrapper.find('FormControl').simulate('change', {
    target: {
      name: 'skill',
      value: 'javascript'
    }
  });
  expect(wrapper.state('skill')).toEqual('javascript');
});

test('handle adding skills', () => {
  wrapper.setState({ skill: 'javascript' });
  wrapper.find('FontAwesome').at(0).simulate('click');
  expect(wrapper.state('skills')).toEqual([...skills, 'javascript']);
  expect(onChange).toHaveBeenCalledWith(wrapper.state('skills'));
});

test('handle removing skills', () => {
  let node = { innerText: 'backbone' };
  wrapper.find('FontAwesome').at(1).simulate('click', {
    target: { parentNode: node }
  });
  expect(wrapper.state('skills')).toEqual([skills[0], skills[1], skills[2]]);
  expect(onChange).toHaveBeenCalledWith(wrapper.state('skills'));
});
