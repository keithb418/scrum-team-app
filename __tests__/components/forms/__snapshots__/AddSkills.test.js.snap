// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddSkills component correctly 1`] = `
<div
  className="add-skills"
>
  <FormGroup
    bsClass="form-group"
    controlId="add-skills"
    style={
      Object {
        "width": "200px",
      }
    }
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Skills
    </ControlLabel>
    <InputGroup
      bsClass="input-group"
    >
      <FormControl
        bsClass="form-control"
        componentClass="input"
        name="skill"
        onChange={[Function]}
        placeholder="Add Skills"
        type="text"
        value=""
      />
      <InputGroupAddon
        bsClass="input-group-addon"
      >
        <FontAwesome
          className="left-spacer add-skills-icon"
          name="plus-circle"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </InputGroupAddon>
    </InputGroup>
  </FormGroup>
  <div
    className="skill"
    key="0"
    style={
      Object {
        "display": "inline-block",
        "marginRight": "15px",
      }
    }
  >
    react
    <FontAwesome
      name="minus-circle"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    />
  </div>
  <div
    className="skill"
    key="1"
    style={
      Object {
        "display": "inline-block",
        "marginRight": "15px",
      }
    }
  >
    redux
    <FontAwesome
      name="minus-circle"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    />
  </div>
  <div
    className="skill"
    key="2"
    style={
      Object {
        "display": "inline-block",
        "marginRight": "15px",
      }
    }
  >
    es6
    <FontAwesome
      name="minus-circle"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    />
  </div>
  <div
    className="skill"
    key="3"
    style={
      Object {
        "display": "inline-block",
        "marginRight": "15px",
      }
    }
  >
    backbone
    <FontAwesome
      name="minus-circle"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    />
  </div>
</div>
`;
